src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -563,27 +537,252 @@ internal SafeFileHandle OpenPort(string portName)
+        private delegate int RequestProcessor(SerialStreamIORequest r);
+        private RequestProcessor _processReadDelegate;
+        private RequestProcessor _processWriteDelegate;
+
+        private unsafe int ProcessRead(SerialStreamIORequest r)
+        {
+            Span<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Read(_handle, bufPtr, buff.Length);
+
+                if (numBytes < 0)
+                {
+                    Interop.ErrorInfo lastError = Interop.Sys.GetLastErrorInfo();
+
+                    // ignore EWOULDBLOCK since we handle timeout elsewhere
+                    if (lastError.Error != Interop.Error.EWOULDBLOCK)
                     {
+                        r.Complete(Interop.GetIOException(lastError));
                     }
-                    base.Dispose(disposing);
+                }
+                else if (numBytes > 0)
+                {
+                    r.Complete(numBytes);
+                    return numBytes;
+                }
+                else // numBytes == 0
+                {
+                    ThreadPool.QueueUserWorkItem(s => {
+                            var thisRef = (SerialStream)s;
+                            thisRef.DataReceived(thisRef, new SerialDataReceivedEventArgs(SerialData.Eof));
+                        }, this);
+                }
+            }
+
+            return 0;
+        }
+
+        private unsafe int ProcessWrite(SerialStreamIORequest r)
+        {
+            ReadOnlySpan<byte> buff = r.Buffer.Span;
+            fixed (byte* bufPtr = buff)
+            {
+                // assumes dequeue-ing happens on a single thread
+                int numBytes = Interop.Sys.Write(_handle, bufPtr, buff.Length);
+
+                if (numBytes <= 0)
+                {
+                    Interop.ErrorInfo lastError = Interop.Sys.GetLastErrorInfo();
+
+                    // ignore EWOULDBLOCK since we handle timeout elsewhere
+                    // numBytes == 0 means that there might be an error
+                    if (lastError.Error != Interop.Error.SUCCESS && lastError.Error != Interop.Error.EWOULDBLOCK)
+                    {
+                        r.Complete(Interop.GetIOException(lastError));
+                    }
+                }
+                else
+                {
+                    r.ProcessBytes(numBytes);
+
+                    if (r.Buffer.Length == 0)
+                    {
+                        r.Complete();
+                    }
+
+                    return numBytes;
+                }
+            }
+
+            return 0;
+        }
+
+        // returns number of bytes read/written
+        private static int DoIORequest(ConcurrentQueue<SerialStreamIORequest> q, RequestProcessor op)
+        {
+            // assumes dequeue-ing happens on a single thread
```

Comment:
>assumes dequeue-ing happens on a single threa [](start = 15, length = 45)

why this assumption?

Reply:
This is only called in the IOLoop which is a single thread. If that wasn't true then TryDequeue could dequeue wrong item than we peeked